request-manager/spec/node-check.go

```diff
@@ -0,0 +1,498 @@
+		}
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type NonconditionalNoEqNodeCheck struct{}
+
+/* Nononditional nodes may not specify `eq`. */
+func (check NonconditionalNoEqNodeCheck) CheckNode(sequenceName string, node NodeSpec) error {
+	if !node.IsConditional() {
+		if len(node.Eq) != 0 {
+			eq := fmt.Sprintf("%v", node.Eq)
+			return InvalidValueError{sequenceName, &node.Name, "eq", []string{eq}, "no value; noncoditional nodes may not specify eq"}
+		}
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type RetryIfRetryWaitNodeCheck struct{}
+
+/* If `retryWait` is set, `retry` must be set (nonzero). */
+func (check RetryIfRetryWaitNodeCheck) CheckNode(sequenceName string, node NodeSpec) error {
+	if node.RetryWait != "" && node.Retry == 0 {
+		return MissingValueError{sequenceName, &node.Name, "retry", "required when `retryWait` field set"}
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type ValidRetryWaitNodeCheck struct{}
+
+/* `retryWait` should be a valid duration. */
+func (check ValidRetryWaitNodeCheck) CheckNode(sequenceName string, node NodeSpec) error {
+	if node.RetryWait != "" {
+		if _, err := time.ParseDuration(node.RetryWait); err != nil {
+			return InvalidValueError{sequenceName, &node.Name, "retryWait", []string{node.RetryWait}, "valid duration string"}
+		}
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type RequiredArgsProvidedNodeCheck struct {
+	specs Specs
+}
+
+/* Sequence and conditional nodes must be provided with their required args. */
+func (check RequiredArgsProvidedNodeCheck) CheckNode(sequenceName string, node NodeSpec) error {
+	// List of sequences to check
+	// (This will also include jobs, but we'll ignore them later)
+	var sequences []string
+	if node.NodeType != nil {
+		sequences = []string{*node.NodeType}
+	}
+	for _, seq := range node.Eq {
+		sequences = append(sequences, seq)
+	}
+
+	// Set of all (declared) inputs to a node
+	var declaredArgs = map[string]bool{}
+	for _, nodeArg := range node.Args {
+		if nodeArg.Expected != nil {
+			declaredArgs[*nodeArg.Expected] = true
+		}
+	}
+	for _, each := range node.Each {
+		p := strings.Split(each, ":")
+		if len(p) != 2 {
+			continue
+		}
+		declaredArgs[p[1]] = true
+	}
+
+	// Check that all required args are present
+	values := map[string]bool{}
+	for _, sequence := range sequences {
+		seq, ok := check.specs.Sequences[sequence]
+		if !ok {
+			continue
+		}
+		for _, reqArg := range seq.Args.Required {
+			if reqArg.Name == nil {
+				continue
+			}
+			if _, ok = declaredArgs[*reqArg.Name]; !ok {
+				values[*reqArg.Name] = true
+			}
+		}
+	}
+
+	if len(values) > 0 {
+		explanation := fmt.Sprintf("required arg(s) \"%s\" not given", strings.Join(stringSetToArray(values), ", "))
```

Comment:
It's not clear from this message that these are required args for a sequence called by this node — can you fix that? Extra points it you can specify which sequence it is that's missing an arg for conditional nodes which can call multiple sequences.